client/cli/src/runner.rs

```diff
@@ -259,24 +264,25 @@ impl<C: SubstrateCli> Runner<C> {
 		}
 	}
 
-	fn run_service_until_exit<T, F>(mut self, service_builder: F) -> Result<()>
-	where
-		F: FnOnce(Configuration) -> std::result::Result<T, sc_service::error::Error>,
-		T: AbstractService + Unpin,
-	{
-		let service = service_builder(self.config)?;
-
-		// we eagerly drop the service so that the internal exit future is fired,
-		// but we need to keep holding a reference to the global telemetry guard
-		// and drop the runtime first.
-		let _telemetry = service.telemetry();
-
-		// we hold a reference to the base path so if the base path is a temporary directory it will
-		// not be deleted before the tokio runtime finish to clean up
-		let _base_path = service.base_path();
+	fn run_service_until_exit(
+		mut self,
+		initialise: impl FnOnce(Configuration) -> sc_service::error::Result<KeepAliveChainComponents>,
+	) -> Result<()> {
+		let KeepAliveChainComponents {
+			task_manager,
+			// we eagerly drop the service so that the internal exit future is fired,
+			// but we need to keep holding a reference to the global telemetry guard
+			// and drop the runtime first.
+			telemetry: _telemetry,
+			// we hold a reference to the base path so if the base path is a temporary directory it will
+			// not be deleted before the tokio runtime finish to clean up
+			base_path: _base_path,
+			rpc: _rpc,
+			..
```

Comment:
(I don't think you would want to have `..` here as it will drop anything added later. It's probably better to keep that fully explicit so compilation would fail if someone wants to add something. Otherwise if defies the very purpose of `KeepAliveChainComponents`)

Reply:
Oh yeah, that was accidentally left there.